pkg/router/template/router.go

```diff
@@ -311,31 +323,174 @@ func (r *templateRouter) readState() error {
 	return json.Unmarshal(data, &r.state)
 }
 
-// Commit applies the changes made to the router configuration - persists
+// Define a type for the pluggable commit function
+type CommitFunc func() error
+
+// This should ONLY be used for external testing hooks
+func (r *templateRouter) SetCommitFunc(commitFunc CommitFunc) {
+	r.commitFunc = commitFunc
+}
+
+// Commit calls the realCommit worker with a flag saying that it has
+// synced.  This should be called only when the router state is
+// consistent.  realCommit() below can be called on initial router
+// load to load the haproxy without a full state sync.
+func (r *templateRouter) Commit() {
+	r.realCommit(true)
+}
+
+// realCommit applies the changes made to the router configuration - persists
 // the state and refresh the backend. This is all done in the background
 // so that we can rate limit + coalesce multiple changes.
+//
+// The hasSynced argument specifies whether the state is fully synchronized.
+//
 // Note: If this is changed FakeCommit() in fake.go should also be updated
-func (r *templateRouter) Commit() {
+func (r *templateRouter) realCommit(hasSynced bool) {
```

Comment:
Wow.  

So this is incredibly complicated and really needs to be abstracted into something smaller and more well tested on its own.  The original point of the "commit wrapper" was to have a function that you invoke that does the right thing, and then the rest of it is ignored.  This has now become entangled with the template router.

Please separate all of this logic out into three pieces - the signal (from the rest of the template router to reload), the rate limiter / tracker (the bit that says when an actual commit should go), and the underlying commit action that controls how fast it happens.

I.e.:

1. router controller signals (changes happened)
2. a rate limiter / commit controller decides whether to pass it on (this code)
3. the underlying commit.

There really cannot be a connection between them, and the code here is too complicated for me to be confident it works.

Reply:
@liggitt can you help suggest here?  i think of this as controller model - several caches:

informer cache -> feeding secondary cache -> controller loop that reads secondary cache and needs to sync up with secondary cache

Reply:
I think I have worked out a possible solution.